src/data.rs

```diff
@@ -94,19 +94,43 @@ impl Token {
 
     pub async fn create_file_writer(&self, name: &str) -> Result<File, String> {
         self.create_referenced_directory().await;
+
+        /* check for finished file name collision */
+        if Path::new(&self.get_final_file_name(name)).exists().await {
+            return Err("file already exists".to_owned());
+        }
```

Comment:
technically, this can race

* another thread is uploading `file` into `file$.partial`
* we check that `file` doesn't exist, and continue
* the other thread finishes and renames `file$.partial` to `file`
* we create new `file$.partial`, it succeeds because it doesn't exist, but we'll fail to rename it in the end.

I guess it doesn't really matter, but I think checking that the filename is free after creating the partial file should be enough to fix it.

Reply:
Thanks, good catch. Fixed.